src/zenml/entrypoints/step_entrypoint_configuration.py

```diff
@@ -299,26 +307,45 @@ def get_entrypoint_arguments(
         # in some docker container entrypoint) use the ZenML constant that
         # points to the actual user main module. If not resolve the `__main__`
         # module to something that can be imported during entrypoint execution.
-        main_module_source = constants.USER_MAIN_MODULE
-        if not main_module_source:
-            main_module_source = source_utils.get_module_source_from_module(
+        main_module_source = (
+            constants.USER_MAIN_MODULE
+            or source_utils.get_module_source_from_module(
                 sys.modules["__main__"]
             )
+        )
 
-        # Get an importable source of the step class. If the step is defined in
-        # the `__main__` module, replace it with the main module source
-        # computed above.
-        step_module = step.__module__
-        if step_module == "__main__":
-            step_module = main_module_source
-        step_source = f"{step_module}.{step.name}"
+        def _resolve_class(class_: Type[Any]) -> str:
+            """Resolves the input class in a way that it is importable inside
+            the entrypoint.
+            """
+            source = source_utils.resolve_class(class_)
+            module_source, class_source = source.rsplit(".", 1)
+            if module_source == "__main__":
+                # If the class is defined inside the `__main__` module it will
+                # not be importable in the entrypoint (The `__main__` module
+                # there will be the actual entrypoint module). We therefore
+                # replace the module source by the importable main module
+                # source computed above.
+                module_source = main_module_source
+
+            return f"{module_source}.{class_source}"
```

Comment:
I would recommend moving this to the `zenml.utils.source_utils` module and re-factoring it as a more generic function, to be used elsewhere in the code.
The reason is that our module/class-to-file resolution logic is already really really complicated (the sheer number of functions and options in the source_utils module by itself is overwhelming), but at least this way we have all the "heavy stuff" co-located in the same module, so if we decide in the future to simplify it (I have a strong feeling it can be simplified), we don't have to gather the necessary pieces from all over.

Reply:
Absolutely agree with both you general statement regarding `source_utils` as well as the place where this function should be in. It would then have to recompute the `main_module_source` each time, right? Because that's used from the outside for now